### ISSUE TITLE
High CPU usage under WSL even when vscode is idle

### ISSUE THREAD

**Issue by tusharsnx** [microsoft/vscode#222854]
Type: <b>Bug</b>

1. Open vscode with wsl:
```sh
(wsl)$ code .
```
2. Use `htop` or similar tool to monitor CPU usage when vscode is idle or minimized to the taskbar.

Expected:
CPU usage for vscode's remote server should stay around 0% at idle conditions.

Actual:
CPU usage stays between 2-10 %

VS Code version: Code 1.91.1 (f1e16e1e6214d7c44d078b1f0607b2388f29d729, 2024-07-09T22:06:49.809Z)
OS version: Windows_NT x64 10.0.26100
Modes:
Remote OS version: Linux x64 6.6.36.3-microsoft-standard-WSL2

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|11th Gen Intel(R) Core(TM) i5-11300H @ 3.10GHz (8 x 3110)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|15.79GB (6.69GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|WSL: Ubuntu-22.04|
|OS|Linux x64 6.6.36.3-microsoft-standard-WSL2|
|CPUs|11th Gen Intel(R) Core(TM) i5-11300H @ 3.10GHz (8 x 0)|
|Memory (System)|7.65GB (6.82GB free)|
|VM|0%|
</details><details><summary>Extensions (10)</summary>

Extension|Author (truncated)|Version
---|---|---
remote-wsl|ms-|0.88.2
material-icon-theme|PKi|5.6.0
vim|vsc|1.27.3
markdown-checkbox|bie|0.4.0
gitlens|eam|15.2.0
prettier-vscode|esb|10.4.0
hexeditor|ms-|1.10.0
advanced-new-file|pat|1.2.2
vscode-fileutils|sle|3.10.3
even-better-toml|tam|0.19.2

</details>

**Comment by tusharsnx** [comment]
I had earlier opened the issue at #222412 where I provided more in-depth issue description. I'm just going to put the same description here.

### Issue Description

I recently noticed high CPU usage and faster battery drain (intel cpu) when working with any project under WSL. To find out what's taking so much cpu time (even when I'm doing nothing, apart from staring at the screen), I used `htop` to list processes in the order of cpu usage:

![Screenshot 2024-07-21 190742](https://github.com/user-attachments/assets/0628ff78-4188-4be9-9f2d-5c6063bd525c)

Looked like vscode's remote server running inside wsl is taking CPU. So, I went a little further to profile the running instance of the remote server and I found:

![Screenshot 2024-07-21 190858](https://github.com/user-attachments/assets/fea1be5e-5755-492d-97f9-0731c7096d32)

![Screenshot 2024-07-21 191001](https://github.com/user-attachments/assets/54b2af48-5926-46c5-80a8-9f023264c138)

`I` from `extensionHostProcess.js` is being called at a regular interval with a small delay (~1000ms). `I`:

![Screenshot 2024-07-21 191040](https://github.com/user-attachments/assets/07bf8e9a-273b-4f00-8a30-d6ff684a2b5e)


I was able to trace `I` back to [findCandidatePorts()](https://github.com/microsoft/vscode/blob/0eec43797f37c7c66e400f0a7a5f8f15e25caddc/src/vs/workbench/api/node/extHostTunnelService.ts#L243):

![Screenshot 2024-07-21 191217](https://github.com/user-attachments/assets/2373095e-cd81-4d91-b5d9-3b632f899ef6)

And I finally found where it is being called repeatedly:

![Screenshot 2024-07-21 191329](https://github.com/user-attachments/assets/233aafa5-30e0-4f42-af10-5b8087a669b3)


Now, why is it a bug? Because I've disabled all `remote.autoForwardPorts` related settings in my config, yet the tunnel monitoring service is running in the background. I expect the service to not be running when `remote.autoForwardPorts` or related configs are disabled. The service is currently always running.

**Comment by aeschli** [comment]
@tusharsnx Wow, great job, thanks for tracking down the problem!

**Comment by alexr00** [comment]
@connor4312 I'm having trouble understanding this:

https://github.com/microsoft/vscode/blob/c8c55ab201ba6946d7edfd15519eb86d83d0b869/src/vs/workbench/api/common/extHostExtensionService.ts#L901-L904

Is WSL going to end up being a `ManagedResolvedAuthority`? I don't think we actually want to show the ports view in WSL even if the Tunnels extension isn't installed.

**Comment by connor4312** [comment]
I agree. For non-managed remote authorities created just via the `remoteAddress` in the workbench, how did we determine whether or not to show the ports view? We should apply the same logic here...

**Comment by alexr00** [comment]
Looks like we also show the ports view when there's an extension with an `onTunnel` actiation event:

https://github.com/microsoft/vscode/blob/35af2fb2086606481fd5f77976748dcc15304176/src/vs/workbench/services/remote/common/tunnelModel.ts#L510-L513

Is this perhaps no longer needed with the work you did for tunnels @connor4312?

**Comment by alexr00** [comment]
Looks like this was actually a feature: https://github.com/microsoft/vscode-remote-release/issues/4923

You can set `"remote.autoForwardPortsSource": "output"` to disable this.

I'm also going to try to improve the description of `remote.autoForwardPorts` and `remote.autoForwardPortsSource`.

**Comment by alexr00** [comment]
To verify:
1. Read the descriptions of `remote.autoForwardPorts` and `remote.autoForwardPortsSource`. Verify that you can understand what the settings do and how they're related. 
2. Set `remote.autoForwardPortsSource`. to `process` or `hybrid` if it isn't already.
3. Using the WSL extension, connect to a WSL remote. 
4. Set the log level to Trace for the Remote Extension Host.
5. Open the output for the Remote Extension Host. Verify that you see a steady stream of log lines that say "candidate" coming in. This means that port discovery is on.
6. Set `remote.autoForwardPortsSource` to `output` and  `remote.autoForwardPorts`  to `false`. Reload.
7. Make sure that the log level of the Remote Extension Host is still set to Trace
8. Open the output for the Remote Extension Host. Verify that you DO NOT see a steady stream of log lines that say "candidate" coming in.

**Comment by alexr00** [comment]
@rzhao271 I can't repro this. Do you have an old version of WSL installed?

**Comment by aeschli** [comment]
I verified that on my machine the [verification steps](https://github.com/microsoft/vscode/issues/222854#issuecomment-2376225009) work.

**Comment by tusharsnx** [comment]
I haven't checked with Insider, but in stable I can confirm 'candidate' logs are not visible after Step 6 from the verification steps.

Thanks.

I also want to point out that it's not clear from the description that there's a process monitoring service running in the background, and that it can be disabled using `remote.autoForwardPortsSource`. Should the default option be `output` instead of  `process`?

**Comment by alexr00** [comment]
Thanks for trying it out @tusharsnx. If we set it to `output`, then we might end up not showing any ports in the Ports view. In insiders, I updated the text of the `source` setting:

Sets the source from which ports are automatically forwarded when [Remote: Auto Forward Ports](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) is true. When [Remote: Auto Forward Ports](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) is false, [Remote: Auto Forward Ports Source](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) will be used to find information about ports that have already been forwarded. On Windows and macOS remotes, the process and hybrid options have no effect and output will be used.

It still doesn't explicitly say there will be process monitoring, but it should be clearer.
